### PR TITLE
add generated weight info for pallet-collective

### PR DESCRIPTION
**Pull request by apopiak**
polkadot companion: https://github.com/paritytech/polkadot/pull/1524

This PR uses generated `WeightInfo` to determine weights for the dispatchables in the collective pallet. It also adds a new benchmark for `disapprove_proposal`.

| extrinsic               | old                                                                   | new                                                                   | notes                            |
|-------------------------|-----------------------------------------------------------------------|-----------------------------------------------------------------------|----------------------------------|
| set_members             | 0 + 21_000_000 * M + 110_000 * N + 27_000_000 * P + 2r + Pr + 2w + Pw | 0 + 21_040_000 * M + 173_000 * N + 31_595_000 * P + 2r + Pr + 2w + Pw | good                             |
| execute                 | 23_000_000 + 4_000 * B + 120_000 * M + 1R + Proposal                  | 43_359_000 + 4_000 * B + 123_000 * M + 1R                             | Proposal weight added to the attribute tag.               |
| propose_execute         | 29_000_000 + 3_000 * B + 220_000 * M + 2R + Proposal                  | 54_134_000 + 4_000 * B + 239_000 * M + 2R                             | proposal weight added to the attribute tag       |
| propose_proposed        | 50_000_000 + 6_000 * B + 110_000 * M + 510_000 * P + 4R + 4W          | 90_650_000 + 5_000 * B + 152_000 * M + 970_000 * P + 4R + 4W          |                                  |
| vote                    | 30_000_000 + 500_000 * M + 1R + 1W                                    | 74_460_000 + 290_000 * M + 2R + 1W                                    | Missed read on `voting` |
| close_early_disapproved | 37_210_000 + 2_000 * B + 239_000 * M + 466_000 * P                    | 86_360_000 + 232_000 * M + 954_000 * P + 3R + 3W                      | DB good                              |
| close_early_approved    | 50_820_000 + 8_000 * B + 211_000 * M + 478_000 * P                    | 123_653_000 + 1_000 * B + 287_000 * M + 920_000 * P + 4R + 3W         | DB good                              |
| close_disapproved       | 51_080_000 + 3_000 * B + 224_000 * M + 475_000 * P                    | 95_395_000 + 236_000 * M + 965_000 * P + 4R + 3W                      | DB good                              |
| close_approved          | 65_950_000 + 5_000 * B + 226_000 * M + 487_000 * P                    | 135_284_000 + 4_000 * B + 218_000 * M + 951_000 * P + 5R + 3W         | DB good                              |
| disapprove_proposal     | 490_000 * P + 1R + 3W                                                 | 50_500_000 + 966_000 * P + 1R + 3W                                    | Big change? Base?                |

### CODE REVIEW

File: frame/collective/src/lib.rs [hunk 1]
```diff
@@ -666,12 +566,17 @@ decl_module! {
...
 		#[weight = (
 			weight_for::close::<T, I>(
-				MAX_MEMBERS.into(), // `M`
+				*length_bound, // B
+				MAX_MEMBERS, // `M`
 				*proposal_weight_bound, // `P1`
-				T::MaxProposals::get().into(), // `P2`
-				*length_bound as Weight, // B
+				T::MaxProposals::get(), // `P2`
 			),
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:580]: actually I would be in favor of writing everything here, as you don't reuse close function anyway:
```rust
		#[weight = (
			{
				let b = *length_bound;
				let m = MAX_MEMBERS;
				let p1 = *proposal_weight_bound;
				let p2 = T::MaxProposals::get();
				T::WeightInfo::close_early_approved(b, m, p2)
					.max(T::WeightInfo::close_early_disapproved(m, p2))
					.max(T::WeightInfo::close_approved(b, m, p2))
					.max(T::WeightInfo::close_disapproved(m, p2))
					.saturating_add(p1)
			},
			DispatchClass::Operational
		)]
```
But I'm ok anyway

File: frame/collective/src/lib.rs [hunk 2]
```diff
@@ -78,26 +78,28 @@ pub const MAX_MEMBERS: MemberCount = 100;
...
-	fn close_disapproved(m: u32, p: u32, b: u32, ) -> Weight;
-	fn close_approved(m: u32, p: u32, b: u32, ) -> Weight;
+	fn close_early_disapproved(m: u32, p: u32, ) -> Weight;
+	fn close_early_approved(b: u32, m: u32, p: u32, ) -> Weight;
+	fn close_disapproved(m: u32, p: u32, ) -> Weight;
+	fn close_approved(b: u32, m: u32, p: u32, ) -> Weight;
+	fn disapprove_proposal(p: u32, ) -> Weight;
 }
 
 impl WeightInfo for () {
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:92]: same as asked by shawn in democracy https://github.com/paritytech/substrate/pull/6783/files better to default to the substrate runtime weight. it avoid breaking change, still not accurate weight in general case but slightly better.

File: bin/node/runtime/src/weights/pallet_collective.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+			.saturating_add(DbWeight::get().reads(4 as Weight))
+			.saturating_add(DbWeight::get().writes(4 as Weight))
+	}
+	fn vote(m: u32, ) -> Weight {
+		(46165000 as Weight)
+			.saturating_add((264000 as Weight).saturating_mul(m as Weight))
+			.saturating_add(DbWeight::get().reads(2 as Weight))
+			.saturating_add(DbWeight::get().writes(1 as Weight))
+	}
+	// WARNING! Some components were not used: ["b"]
```
Comment (thiolliere) [inline bin/node/runtime/src/weights/pallet_collective.rs:58]: As far as I see close_early_disapproved is not O(b) anyway so benchmark should remove it no ?

File: frame/collective/src/lib.rs [hunk 4]
```diff
@@ -591,8 +488,11 @@ decl_module! {
 		///   - 1 storage mutation `Voting` (codec `O(M)`)
 		/// - 1 event
 		/// # </weight>
+		// NOTE: surprising difference between old manual weight and new generated weight:
+		// (49155000 as Weight)
+		//	.saturating_add((295000 as Weight).saturating_mul(m as Weight))
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:493]: | before | after |
|------|-----|
| 30_000_000 + 500_000 * m | 46_165_000 + 264_000*m |

surprising but not too bad seems like the analysis moved some cost related to m to the constant cost 🤔

File: frame/collective/src/lib.rs [hunk 5]
```diff
@@ -666,12 +566,17 @@ decl_module! {
 		///  - any mutations done while executing `proposal` (`P1`)
 		/// - up to 3 events
 		/// # </weight>
+		// NOTE: surprising difference between old manual weight and new generated weight:
+		// early disapproved: 37.21 vs 47.412
+		// early approved:    50.82 vs 69.466 and B * 0.008 vs 0.004 * b
+		// disapproved:       B * 0.003 vs 0 * b
+		// approved:          65.95 vs 77.033
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:573]: its interesting that most of the benchmark go increasing. I wonder why.

about disapproved I think it is correct to have no influence of b anymore.

**Comment by thiolliere** [comment]
I guess you checked that read/writes were consistant, about weight changes, it doesn't look to have change so much, maybe having an array of comparison would help

Otherwise looks good to me

**Comment by apopiak** [comment]
> I guess you checked that read/writes were consistant [...]

yeah, the new weights seemed more accurate, even :-D

File: frame/collective/src/lib.rs (hunk 2 above)
Comment (apopiak) [reply frame/collective/src/lib.rs:92]: so I'll also remove them

Comment (thiolliere) [reply frame/collective/src/lib.rs]: I think what was agreed was to implement it with a non breaking change default_weight (i.e. current polkadot weights)
Thus you should just copy the file from node/runtime/weight to here and modify it a bit so it implements for (). You can see democracy in master to take example, Shawn approved it.

Comment (apopiak) [reply frame/collective/src/lib.rs]: democracy on master does not provide a `()` implementation AFAICT?

Comment (thiolliere) [reply frame/collective/src/lib.rs]: it is done here https://github.com/paritytech/substrate/blob/master/frame/democracy/src/default_weight.rs

**Comment by gavofyork** [comment]
@shawntabrizi final check would be nice.

File: frame/collective/src/lib.rs [hunk 6]
```diff
@@ -667,12 +533,17 @@ decl_module! {
...
+			{
+				let b = *length_bound;
+				let m = MAX_MEMBERS;
+				let p1 = *proposal_weight_bound;
+				let p2 = T::MaxProposals::get();
+				T::WeightInfo::close_early_approved(b, m, p2)
+					.max(T::WeightInfo::close_early_disapproved(m, p2))
+					.max(T::WeightInfo::close_approved(b, m, p2))
+					.max(T::WeightInfo::close_disapproved(m, p2))
+					.saturating_add(p1)
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:545]: I guess p1 shouldn't be added in close_disapproved nor in close_early_disapproved, but the final result is same

**Review by thiolliere — APPROVED** [review]
I compared weight before/after, overall value are similar except:
* close: before: 4 reads, 2 writes; now: 5/3 (in close_approved) 

maybe there was some miss before ?

File: frame/collective/src/lib.rs [hunk 7]
```diff
@@ -453,11 +319,10 @@ decl_module! {
 		/// - 1 event
 		/// # </weight>
 		#[weight = (
-			weight_for::execute::<T, I>(
-				MAX_MEMBERS.into(),
-				proposal.get_dispatch_info().weight,
-				*length_bound as Weight,
-			),
+			T::WeightInfo::execute(
+				*length_bound as u32, // B
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:323]: in case length_bound is somehow moved to u64, this is safer
```suggestion
				*length_bound, // B
```

**Review by shawntabrizi — APPROVED** [review]

File: bin/node/runtime/src/weights/pallet_collective.rs [hunk 8]
```diff
@@ -13,83 +13,84 @@
...
+			.saturating_add((218000 as Weight).saturating_mul(m as Weight))
+			.saturating_add((951000 as Weight).saturating_mul(p as Weight))
 			.saturating_add(DbWeight::get().reads(5 as Weight))
 			.saturating_add(DbWeight::get().writes(3 as Weight))
 	}
 	fn disapprove_proposal(p: u32, ) -> Weight {
-		(0 as Weight)
-			.saturating_add((490000 as Weight).saturating_mul(p as Weight))
+		(50500000 as Weight)
+			.saturating_add((966000 as Weight).saturating_mul(p as Weight))
```
Comment (apopiak) [inline bin/node/runtime/src/weights/pallet_collective.rs:93]: that's a lot more weight than I expected, what do you think where it comes from?

**Review by apopiak — COMMENTED** [review]
LGTM

**Comment by apopiak** [comment]
bot merge

**Pull request merged**